Remove Python2 support

jnewbery opened the pull request:
Following discussion here: https://github.com/bitcoin/bitcoin/pull/11843#issuecomment-351033742

It's easier for maintainers if all python tools/scripts support only a single version of Python. There are only a few scripts that aren't explicitly python3 at this point, so this PR changes those remaining scripts to explicitly require python3.

jnewbery committed:
Change all python files to use Python3

practicalswift commented:
Strong concept ACK!

Pros:
* Explicit requirements are better than implicit requirements
* Less work required to maintain the Python scripts
* Avoids potential bugs due to subtle Python 2/3 incompatibilities

Cons:
* ?

MarcoFalke commented:
ACK for devtools. Not sure about unit tests and user facing scripts such as rpcauth.py, see also discussion in #11433

jnewbery commented:
Marco, your comment from #11433:

> I am not against deprecating py2, but we should do it properly such that ./configure fails early and warns about it. Also it should be mentioned in the release notes.

Would you support this PR for all Python files including user-facing scripts and tests if it met those conditions?

droark commented:
Concept ACK. I'm not aware of any major projects still on Py2 and intending to stay there. AFAIK, just about everything that's still Py2 is due to legacy and/or manpower issues. Here's how I'd like to see things handled.

- Make the decision to go Py3-only.
- Add something in release notes for the next major version (0.16 or 0.17, I assume) indicating that Py2 will no longer be usable in the following version.
- Make the changes, and force any config/make materials reliant on Python to ensure that Py3 is present.

MarcoFalke commented:
@jnewbery I'd be -0 on that. The reason that the functional tests run at 3.4+ only, is because it would be nearly impossible to have them compatible with both 2.7 and 3.4. However, the python code used for the unit tests and the few scripts mean for users is rather condensed and hopefully not much refactored in the future.
Note that the loose scripts are not (regularly) tested on py2 nor py3, so just declaring that they no longer function with py2 isn't a huge improvement. (Or making them merly incompatible with py2 isn't a huge improvement either, imo). As long as they are not rewritten in a large-ish way, I don't see why we should change the dependency requirements...

laanwj commented:
ACK on the idea to go Py3-only long-term. There's a burden on supporting multiple versions of python and that burden will only become larger as Python 3 progresses, by locking us out of new features and better ways to do things.

(this is a similar argument as that for going to C++11)

The only problematic case are scripts that depend on libraries that only support Python2. These are rare, but I vaguely remember some of the macdeploy scripts do, so please test those carefully.

> As long as they are not rewritten in a large-ish way, I don't see why we should change the dependency requirements...

I also agree with this, though. There's no urgent reason to make this change at once, certainly for scripts that are hardly touched.

droark commented:
@laanwj - I haven't tried the scripts yet but I think macdeploy has a good shot at running on Py3 now. It looks like there are three external packages: [biplist](https://github.com/al45tair/ds_store/commits/master), [ds_store](https://github.com/al45tair/ds_store/commits/master), and [mac_alias](https://github.com/al45tair/mac_alias/commits/master). The commit lists seem to indicate that there's a reasonable chance all of them have Py3 support, or at least enough to get macdeploy to run. I'll confirm and submit upstream patches if necessary.

fanquake commented:
We are still patching [biplist and mac_alias](https://github.com/bitcoin/bitcoin/tree/master/depends/patches), so you could definitely send those patches upstream.

laanwj commented:
OTOH the situation of having some things work in py2 but not others is
confusing. It doesn't really benefit anyone, and there's also an overhead
explaining it to contributors. That's one argument for doing this at once.

On Dec 12, 2017 20:53, "John Newbery" <notifications@github.com> wrote:

Following discussion here: #11843 (comment)
<https://github.com/bitcoin/bitcoin/pull/11843#issuecomment-351033742>

It's easier for maintainers if all python tools/scripts support only a
single version of Python. There are only a few scripts that aren't
explicitly python3 at this point, so this PR changes those remaining
scripts to explicitly require python3.

Still a WIP, and all of these changes need to be reviewed/tested
thoroughly. Looking for high-level feedback on whether this is wanted.
------------------------------
You can view, comment on, or merge this pull request online at:

  https://github.com/bitcoin/bitcoin/pull/11881
Commit Summary

   - Python2 -> Python3

File Changes

   - *M* contrib/devtools/check-doc.py
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-0> (2)
   - *M* contrib/devtools/clang-format-diff.py
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-1> (10)
   - *M* contrib/devtools/optimize-pngs.py
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-2> (26)
   - *M* contrib/devtools/security-check.py
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-3> (13)
   - *M* contrib/devtools/symbol-check.py
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-4> (9)
   - *M* contrib/devtools/update-translations.py
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-5> (3)
   - *M* contrib/macdeploy/custom_dsstore.py
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-6> (3)
   - *M* contrib/macdeploy/macdeployqtplus
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-7> (19)
   - *M* contrib/testgen/gen_base58_test_vectors.py
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-8> (2)
   - *M* share/qt/extract_strings_qt.py
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-9> (3)
   - *M* share/rpcauth/rpcauth.py
   <https://github.com/bitcoin/bitcoin/pull/11881/files#diff-10> (2)

Patch Links:

   - https://github.com/bitcoin/bitcoin/pull/11881.patch
   - https://github.com/bitcoin/bitcoin/pull/11881.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/bitcoin/bitcoin/pull/11881>, or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAHutk62BIlCBa3F5D6YqvuTRD_y6hwVks5s_tmpgaJpZM4Q_imm>
.

droark commented:
@fanquake - Yep, mac_alias doesn't have those fixes. I'll submit upstream. I'll also try with biplist but I don't know how that'll go, since it's not a Py3 fix.

droark commented:
Have submitted PRs for [biplist](https://bitbucket.org/wooster/biplist/pull-requests/9) (not Py3) and [mac_alias](https://github.com/al45tair/mac_alias/pull/3). It's a start!

ryanofsky commented on contrib/devtools/optimize-pngs.py:
```diff
@@ -37,7 +37,7 @@ def content_hash(filename):
         extension = os.path.splitext(file)[1]
         if extension.lower() == '.png':
-            print("optimizing "+file+"..."),
+            print("optimizing "+file+"...")
```
Trailing comma in python2 outputs space instead of newline. I think equivalent in python3 would be 

```python
print("optimizing {}...".format(file), end=' ')
```

jnewbery replied:
thanks. Fixed

MarcoFalke replied:
Was this fixed?

ryanofsky commented on contrib/devtools/optimize-pngs.py:
```diff
@@ -46,33 +46,33 @@ def content_hash(filename):
             fileMetaMap['psize'] = os.path.getsize(file_path)
             outputArray.append(fileMetaMap)
-            print("done\n"),
+            print("done\n")
```
equivalent would be `print("done")`

jnewbery replied:
thanks. Fixed

MarcoFalke replied:
Same here

ryanofsky commented on contrib/macdeploy/macdeployqtplus:
```diff
@@ -719,14 +718,14 @@ if config.dmg is not None:
                 cmd = kwargs.get("args")
                 if cmd is None:
                     cmd = popenargs[0]
-                raise CalledProcessError(retcode, cmd)
```
This was just broken before, I guess? I don't see how it could have worked.

jnewbery replied:
indeed

ryanofsky reviewed (commented):
utACK bbe67bdd82925cc72a868bb3de1896e61cae79db. Since this marked WIP, it might be good to update the PR description to say what other changes are underway or planned.

laanwj commented:
Needs rebase, I think we should do this for 0.17.

droark commented:
Just for cross-referencing, the mac_alias patch made it into the mainline and was removed from Core [here](https://github.com/bitcoin/bitcoin/pull/12417). I pinged the biplist maintainer. He said he'd been busy and would take a look at my PR eventually.

practicalswift commented:
@laanwj Excellent! Let's get rid of Python 2!

droark commented:
Hi. Andrew (the biplist maintainer) reviewed the Core patch and rejected it. Here's his rationale. Any thoughts? I admit that I'm not super familiar with the biplist codebase.

> I looked into this, and it looks like v1.0.2 added sorting of sets and dictionary keys when writing, which is the change needed for reproducible builds. The patch from Bitcoin Core targets biplist v0.9, which didn't have this sorting.

> The below two changes from the Bitcoin Core patch seem to have any effect when writing plists, AFAICT (please correct me if I am wrong). One is wrapping the root dict in a hashable wrapper, so sorting the keys doesn't matter. The other is for computing some values in the file trailer related to how many unique objects there are in the file and how many bytes it takes to represent the offsets to those objects within the file.

> That said, this is under-tested, so I [added a test](https://bitbucket.org/wooster/biplist/commits/4c3d0c94132621188ca4c1fa9bda3b2d4d52ab82) for what I expect the output to be.

> Thanks!

jnewbery commented:
Rebased and addressed @ryanofsky's review comments.

> Since this marked WIP, it might be good to update the PR description to say what other changes are underway or planned.

This is marked as WIP since I was looking for initial feedback and haven't tested or carefully reviewed my suggested changes. Since there seems to be some enthusiasm for removing python2 support, I'll clean this PR up.

jtimon commented:
Concept ACK

eklitzke commented:
Concept ACK.

wrt the comments from @droark, it's easy to wrap iterator types using `sorted()` or types like `collections.OrderedDict()`. If reproducibility is an issue then we should handle that, but that's possible to achieve in Py3 (and if Py2 code relies on it, that code is broken anyway).

ryanofsky commented:
utACK 4f3238250bc01eedee7c60b4d6681d3e21c6783c. Changes since last review were porting `check-doc.py`, fixing newlines for prints with trailing commas, updating copyright years.

practicalswift commented:
utACK 4f3238250bc01eedee7c60b4d6681d3e21c6783c

MarcoFalke commented on contrib/macdeploy/macdeployqtplus:
```diff
@@ -203,7 +202,7 @@ def getFrameworks(binaryPath, verbose):
     otoolbin=os.getenv("OTOOL", "otool")
-    otool = subprocess.Popen([otoolbin, "-L", binaryPath], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    otool = subprocess.Popen([otoolbin, "-L", binaryPath], stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
     o_stdout, o_stderr = otool.communicate()
```
You changed this to `universal_newlines`, so the type should be `str`.

Thus, the `decode` a couple of lines down will fail.

jnewbery replied:
Thanks. Fixed

MarcoFalke commented:
* Changed title-tag.
* Needs feedback addressed.

jnewbery commented:
Thanks all for review. Will address feedback next week.

MarcoFalke commented on contrib/devtools/check-doc.py:
```diff
@@ -33,12 +33,12 @@ def main():
   args_need_doc = args_used.difference(args_docd)
```
One line above this:

You are switching this file to python3, so `check_output` will return byte-strings instead of text strings. Thus, you can not apply a regex on them.

ryanofsky replied:
Should be fixable if you switch the patterns to bytestrings `b'pattern'`

jnewbery commented:
I've gone through and reviewed/tested my changes, and I think they're good. Ready for wider review.

The only python file that is not now explicitly Python3 is `gen_base58_test_vectors.py`. It would require some substantial modifications to work in Python3 (since it's using string/bytes manipulation which changed completely between Python2 and Python3). That file is not modified or used very much, so I've left it as Python2, adding an explicit note to the doc string saying that it's Python2 only.

MarcoFalke commented on doc/release-notes.md:
```diff
@@ -117,6 +117,11 @@ deprecated in V0.15.1, and has now been removed. Miners should use the
+Python Support
+--------------
+
+Support for Python 2 has been discontinued for all test files and tools.
```
NACK based on my earlier feedback in this pull request.

jnewbery replied:
I disagree. We should remove support for a legacy language which will soon not be maintained, and which may hide subtle bugs due to incompatibilities with the new language.

What do we gain by claiming to support python2?

MarcoFalke replied:
I am not saying we should claim to support python2, but we should not claim that python2 support has been discontinued for "all test files" (specifically *util tests*, since the functional tests are already on python3). I think it is reasonable to support python2 util tests until we need to switch off of it because it would otherwise be a burden to continue to support python2. I don't see why we should put effort into making the util tests incompatible with python2 when we might as well leave them supported on a best effort basis.

droark replied:
@MarcoFalke - Honestly, supporting on a "best effort basis" seems like a nightmare waiting to happen. Person A introduces a commit that breaks Py2 compatibility, Person B complains that unit tests are failing, Persons C-? have to deal with the whole thing. I think having 2 & 3 supported is good for transition purposes, and making sure that nothing is broken during a switch. Beyond that, I'm with @jnewbery and don't see any real gain from claiming to support Py2. At most, I'd be okay with a statement mentioning that, while Py2 compatibility is best effort, people *must* use Py3 before complaining that unit tests fail.

MarcoFalke replied:
I think it is not worth arguing about that.

maflcko reviewed (commented):
Concept ACK on the diff (excluding change to `doc/release-notes.md`)

Will review after revert.

laanwj commented:
> That file is not modified or used very much, so I've left it as Python2, adding an explicit note to the doc string saying that it's Python2 only.

I might port it over some time - agree it's not relevant to this PR.

MarcoFalke commented:
utACK bc6fdf2d15648a5fc68df8021d9186737de6fe7b

eklitzke committed:
Make base58 python contrib code work with python3

eklitzke commented:
ACKbc6fdf2d15648a5fc68df8021d9186737de6fe7b

@jnewbery @MarcoFalke cherry-pick https://github.com/eklitzke/bitcoin/commit/563b4f5e398ca8824d42fe271d1485eb3934ed41 if you would like to make the base58 code work on python3

jnewbery commented:
Thanks @eklitzke - I've cherry-picked your commit (updating the hashbang and docstring).

practicalswift commented:
utACK 18740586baee546064cba9286e2d681a849ae443

ken2812221 commented:
gitian build fails on security-check.py
<pre>
+ make -j4 -C src check-security
make: Entering directory `/home/ubuntu/build/bitcoin/distsrc-i686-w64-mingw32/src'
Checking binary security...
Traceback (most recent call last):
  File "../contrib/devtools/security-check.py", line 201, in <module>
    if not func(filename):
  File "../contrib/devtools/security-check.py", line 144, in check_PE_DYNAMIC_BASE
    (arch,bits) = get_PE_dll_characteristics(executable)
  File "../contrib/devtools/security-check.py", line 130, in get_PE_dll_characteristics
    for line in stdout.split('\n'):
TypeError: 'str' does not support the buffer interface
make: *** [check-security] Error 1
make: Leaving directory `/home/ubuntu/build/bitcoin/distsrc-i686-w64-mingw32/src'
</pre>

jnewbery commented:
huh. That's confusing. All the other calls to `subprocess.Popen` expect stdout to be bytes, but line 124 expects stdout to be a string. Will open a fixup PR.

jnewbery commented:
@ken2812221 thanks for reporting this. Can you try again with https://github.com/jnewbery/bitcoin/tree/python3_fixup and let me know if it resolves your issue?

ken2812221 commented:
Wait, I think the problem is that gitian build does not use python3

ken2812221 commented:
<pre>
+ make -j4 -C src check-security
make: Entering directory `/home/ubuntu/build/bitcoin/distsrc-i686-w64-mingw32/src'
Checking binary security...
Traceback (most recent call last):
  File "../contrib/devtools/security-check.py", line 201, in <module>
    if not func(filename):
  File "../contrib/devtools/security-check.py", line 144, in check_PE_DYNAMIC_BASE
    (arch,bits) = get_PE_dll_characteristics(executable)
  File "../contrib/devtools/security-check.py", line 133, in get_PE_dll_characteristics
    arch = tokens[1].rstrip(',')
TypeError: 'str' does not support the buffer interface
make: *** [check-security] Error 1
make: Leaving directory `/home/ubuntu/build/bitcoin/distsrc-i686-w64-mingw32/src'
</pre>

ken2812221 commented:
It also fails on building linux
https://gist.github.com/ken2812221/5ef0feec721f48867bbd23b92593e508

jnewbery commented:
Thanks @ken2812221 . I've pushed a new commit. Can you retry?

ken2812221 commented:
<pre>
+ make -j4 -C src check-security
make: Entering directory `/home/ubuntu/build/bitcoin/distsrc-i686-w64-mingw32/src'
Checking binary security...
Traceback (most recent call last):
  File "../contrib/devtools/security-check.py", line 192, in <module>
    etype = identify_executable(filename)
  File "../contrib/devtools/security-check.py", line 182, in identify_executable
    if magic.startswith('MZ'):
TypeError: startswith first arg must be bytes or a tuple of bytes, not str
make: *** [check-security] Error 1
make: Leaving directory `/home/ubuntu/build/bitcoin/distsrc-i686-w64-mingw32/src'
</pre>

jnewbery commented:
New branch pushed. Can you try again?

MarcoFalke commented on contrib/devtools/optimize-pngs.py:
```diff
@@ -37,42 +37,40 @@ def content_hash(filename):
-                pngCrushOutput = subprocess.check_output(
-                        [pngcrush, "-brute", "-ow", "-rem", "gAMA", "-rem", "cHRM", "-rem", "iCCP", "-rem", "sRGB", "-rem", "alla", "-rem", "text", file_path],
-                        stderr=subprocess.STDOUT).rstrip('\n')
+                subprocess.call([pngcrush, "-brute", "-ow", "-rem", "gAMA", "-rem", "cHRM", "-rem", "iCCP", "-rem", "sRGB", "-rem", "alla", "-rem", "text", file_path],
```
Post-merge nit: Should probably use `check_call` instead.

ken2812221 commented:
Building windows is OK, but buiding linux still fail at check-symbols as I post on gist

jnewbery commented:
Please retry check-symbols.py with latest commit.

jnewbery commented:
@ken2812221 , thanks so much for your patience and help here. I've opened a PR to fix symbol-check and security-check: #12829

practicalswift commented:
Goodbye Python 2. Welcome Python 3!
